packages/amplify-provider-awscloudformation/src/utils/consolidate-apigw-policies.ts

```diff
@@ -0,0 +1,271 @@
+
+  Object.keys(apis).forEach(resourceName => {
+    const resource = apis[resourceName];
+    const apiParams = loadApiWithPrivacyParams(context, resourceName, resource);
+
+    if (!apiParams) {
+      return;
+    }
+
+    const api = { ...resource, resourceName, params: apiParams };
+    updateExistingApiCfn(context, api);
+    apiGateways.push(api);
+  });
+
+  if (apiGateways.length === 0) {
+    return {};
+  }
+
+  return createApiGatewayAuthResources(context, stackName, apiGateways);
+}
+
+function createApiGatewayAuthResources(context: $TSContext, stackName: string, apiGateways: $TSAny): $TSObject {
+  const stack = new ApiGatewayAuthStack(undefined, 'Amplify', {
+    description: 'API Gateway policy stack created using Amplify CLI',
+    stackName,
+    apiGateways,
+  });
+  const cfn = stack.toCloudFormation();
+  const { amplify } = context;
+  const { DeploymentBucketName } = amplify.getProjectMeta()?.providers?.[ProviderName] ?? {};
+  const cfnPath = path.join((amplify.pathManager as any).getBackendDirPath(), 'api', `${APIGW_AUTH_STACK_LOGICAL_ID}.json`);
+
+  JSONUtilities.writeJson(cfnPath, cfn);
+
+  return {
+    APIGatewayAuthURL: `https://s3.amazonaws.com/${DeploymentBucketName}/amplify-cfn-templates/${S3_UPLOAD_PATH}`,
+  };
+}
+
+export function loadApiWithPrivacyParams(context: $TSContext, name: string, resource: any): object | undefined {
+  if (resource.providerPlugin !== ProviderName || resource.service !== 'API Gateway') {
+    return;
+  }
+
+  const apiParamsPath = path.join(getResourceDirPath(context, 'api', name), API_PARAMS_FILE);
+  const apiParams: any = JSONUtilities.readJson(apiParamsPath, { throwIfNotExist: false }) ?? {};
+
+  if (apiParams?.privacy?.auth === 0 && apiParams?.privacy?.unauth === 0) {
+    return;
+  }
+
+  return apiParams;
+}
+
+function updateExistingApiCfn(context: $TSContext, api: $TSObject): void {
+  const { resourceName } = api.params;
+  const resourceDir = getResourceDirPath(context, 'api', resourceName);
+  const cfnTemplate = path.join(resourceDir, `${resourceName}-cloudformation-template.json`);
+  const paramsFile = path.join(resourceDir, 'parameters.json');
+  const cfn: any = JSONUtilities.readJson(cfnTemplate, { throwIfNotExist: false }) ?? {};
+  const parameterJson = JSONUtilities.readJson(paramsFile, { throwIfNotExist: false }) ?? {};
+  const parameters = cfn?.Parameters ?? {};
+  const resources = cfn?.Resources ?? {};
+  let modified = false;
+
+  for (const parameterName in parameters) {
+    if (parameterName === AUTH_ROLE_NAME || parameterName === UNAUTH_ROLE_NAME) {
+      delete parameters[parameterName];
+      delete parameterJson[parameterName];
+      modified = true;
+    }
+  }
+
+  // eslint-disable-next-line guard-for-in
+  for (const resourceName in resources) {
+    const resource = resources[resourceName];
+
+    if (resource.Type === 'AWS::IAM::Policy') {
+      const roles = resource?.Properties?.Roles;
+      const roleName = roles?.[0].Ref;
```

Comment:
So this assumes that if roles is defined, then it has at least 1 element. Length check from below could be moved before this to be on the safe side.